hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java

```diff
@@ -58,132 +70,192 @@ public HoodieCommitArchiveLog(HoodieWriteConfig config, FileSystem fs) {
+                        .withFs(fs)
+                        .overBaseCommit("").build();
+            } else {
+                return this.writer;
+            }
+        } catch(InterruptedException | IOException e) {
+            throw new HoodieException("Unable to initialize HoodieLogFormat writer", e);
         }
-        return success & deleteOtherInstants();
     }
 
-    private boolean deleteOtherInstants() {
-        // Delete clean and rollback files
-        List<HoodieInstant> toDelete = getInstantsToDelete().collect(Collectors.toList());
-        if(!toDelete.isEmpty()) {
-            log.info("Deleting actions " + toDelete);
-            return deleteInstants(toDelete);
+    private void close() {
+        try {
+            if(this.writer != null) {
+                this.writer.close();
+            }
+        } catch(IOException e) {
+            throw new HoodieException("Unable to close HoodieLogFormat writer", e);
         }
-        return true;
     }
 
-    private Stream<HoodieInstant> getInstantsToDelete() {
-
-        int maxCommitsToKeep = config.getMaxCommitsToKeep();
-        int minCommitsToKeep = config.getMinCommitsToKeep();
-
-        HoodieTable table = HoodieTable.getHoodieTable(new HoodieTableMetaClient(fs, config.getBasePath(), true), config);
-        HoodieTimeline cleanTimeline = table.getActiveTimeline().getTimelineOfActions(Sets.newHashSet(HoodieTimeline.CLEAN_ACTION));
-        if (!cleanTimeline.empty() && cleanTimeline.countInstants() > maxCommitsToKeep) {
-            // Actually do the commits
-            return cleanTimeline.getInstants().limit(cleanTimeline.countInstants() - minCommitsToKeep);
+    /**
+     * Check if commits need to be archived. If yes, archive commits.
+     */
+    public boolean archiveIfRequired() {
+        try {
+            List<HoodieInstant> instantsToArchive = getInstantsToArchive().collect(Collectors.toList());
+            boolean success = true;
+            if (instantsToArchive.iterator().hasNext()) {
+                this.writer = openWriter();
+                log.info("Archiving instants " + instantsToArchive);
+                archive(instantsToArchive);
+                success = deleteArchivedInstants(instantsToArchive);
+            } else {
+                log.info("No Instants to archive");
+            }
+            return success;
+        } finally {
+            close();
         }
-        return Stream.empty();
     }
 
-    private Stream<HoodieInstant> getCommitsToArchive() {
+    private Stream<HoodieInstant> getInstantsToArchive() {
 
         int maxCommitsToKeep = config.getMaxCommitsToKeep();
         int minCommitsToKeep = config.getMinCommitsToKeep();
 
         HoodieTable table = HoodieTable.getHoodieTable(new HoodieTableMetaClient(fs, config.getBasePath(), true), config);
+
+        HoodieTimeline cleanAndRollbackTimeline = table.getActiveTimeline().getTimelineOfActions(Sets.newHashSet(HoodieTimeline.CLEAN_ACTION,
```

Comment:
Can you unify the timeline of commit, clean and rollback and do the relevant filter and return back the matching instants that should be archived? That will make the code look clean. Instead of computing this seperately and doing a Stream.concat